pkg/storage/replicas_storage.go

```diff
@@ -462,6 +462,52 @@ import (
 //   RangeAppliedState.RaftAppliedIndex to use as an additional upper bound of
 //   what can be truncated.
 //
+//   Addendum: As of https://github.com/cockroachdb/cockroach/pull/80193 the
+//   loosely-coupled raft log truncation, described above, is disabled due to
+//   a performance regression in write-heavy workloads (see comment
+//   https://github.com/cockroachdb/cockroach/issues/78412#issuecomment-1119922463
+//   for conclusion of investigation). The revised plan is to
+//
+//   - Do strongly-coupled truncation in
+//     CanTruncateRaftIfStateMachineIsDurable for a ReplicasStorage
+//     implementation that shares the same engine for the state machine and
+//     raft state. This relies on external code structure for correctness: the
+//     truncation proposal flows through raft, so we have already applied the
+//     state machine changes for the preceding entries. A crash will cause a
+//     suffix of the unsynced changes to be lost, so we
+//     cannot lose the state machine changes while not losing the truncation.
+//
+//     This is the same correctness argument that the code preceding
+//     ReplicasStorage relies on. The separation of the RaftMutationBatch
+//     provided to DoRaftMutation and the MutationBatch provided to
+//     ApplyCommittedBatch is only more formalization of the separation that
+//     already exists: handleRaftReadyRaftMuLocked makes raft changes with one
+//     batch, and replicaAppBatch.ApplyToStateMachine is used to make changes
+//     to the state machine with another batch. This claim is oversimplified
+//     since the latter also does the raft log truncation, and raft changes
+//     for splits and merges, which ReplicasStorage is doing in a different
+//     way, but it does not change the correctness claim.
+//
+//  -  Do loosely-coupled truncation in CanTruncateRaftIfStateMachineIsDurable
+//     for a ReplicasStorage implementation that has different engines for the
+//     state machine and raft state. The RaftLogTruncator (for loosely-coupled
+//     truncation) will be used by ReplicasStorage for this purpose. This will
+//     delay log truncation, but we speculate that this will not be a
+//     performance problem (experiments will be needed to determine the actual
+//     behavior):
+//     - With multiple key-value pairs in a batch, the memtable for the raft
+//       engine will be able to store more than the corresponding state
+//       machine memtable where the key-value pairs get individual entries in
+//       the memtable. This is because of the per-entry overhead. This means
+//       there is a decent probability that the state machine memtable will
+//       start getting flushed before the corresponding raft engine memtable
+//       is flushed. If the flush is fast enough, we would be able to truncate
```

Comment:
Could we also "just" run the raft instance with a larger memtable?

Reply:
This is my thinking as well. With the separate engines, we should be able to configure it specifically for this use-case -- I'm sure there are other tweaks we could make too.

Reply:
This is my thinking as well. With the separate engines, we should be able to configure it specifically for this use-case -- I'm sure there are other tweaks we could make too.